baybe/recommenders/base.py

```diff
@@ -81,23 +88,15 @@ def _select_candidates_and_recommend(
     return rec
 
 
-@define
-class Recommender(ABC):
-    """Abstract base class for all recommenders."""
-
-    # Class variables
-    compatibility: ClassVar[SearchSpaceType]
-    """Class variable describing the search space compatibility."""
+class RecommenderProtocol(Protocol):
+    """Type protocol specifying the interface recommenders need to implement."""
 
-    @abstractmethod
     def recommend(
         self,
         searchspace: SearchSpace,
-        batch_quantity: int = 1,
-        train_x: Optional[pd.DataFrame] = None,
-        train_y: Optional[pd.DataFrame] = None,
-        allow_repeated_recommendations: bool = False,
-        allow_recommending_already_measured: bool = True,
+        batch_quantity: int,
```

Comment:
Absolutely agree, having default values here wouldn't make sense. What do you suggest? Technically, we don't need to define them on the abstract level (protocol / base class) since all of them recommenders will have their specific values in their signatures anyway. But I'm open for ideas.

And BTW: Since `Recommender` is an ABC, I don't think it's crazy that it does not implement the protocol, but I also haven't seen any good practices for that. Have you done some quick search already?

Reply:
yes I would just leave defaults out of the protocol

Haven seen this particular method mentioned, but I found an article that examined Protocols and ABCs as two different ways of achieving roughly similar things. So in one sense it might be weird that we use both, in another I didnt really see big problems or unreadable code so I guess its fine